### ISSUE TITLE
quite a few problems with built-in Notebook support

### ISSUE THREAD

**Issue by lewisl** [julia-vscode/julia-vscode#2638]
There are several inconsistencies between using the notebook support built-in to the Julia vscode extension and IJulia kernels using the Python notebook runtime. And the VS Code Jula kernels seem very unreliable.

Here are some issues I've discovered with the vscode extension Julia notebook support.

1. Notebooks frequently and easily lose their connection to the Julia vscode extension kernel. I can't completely pinpoint the cause because the problem is intermittent. Shifting focus away from VS Code and leaving VS Code idle for sometime will cause a drop when I context switch back to VSCode.   The IJulia kernels seem to stay connected more reliably. I wish I could provide a more deterministic way to reproduce this, but it happens a lot.

2. There are no ways with the extension kernel to manage the kernel.  The menubar commands for restart and interrupt are not present. So, there is no way to restart the kernel after it has been dropped (or a code error has occurred--see next point).  The only way appears to be to close the notebook tab (what VS Code calls an "editor") and re-open it. Not huge, but given how frequently it must be done because of dropped connections, it is a problem.

3. You cannot recover from errors in cells.  If Julia encounters a compile-time or run time error in a cell, resulting in red circled x where the cell completion time would be displayed, then the notebook has hung. You cannot run any other cell because the errored cell did not complete and no other queued cell can run until previously scheduled cells run. Since you cannot interrupt or restart the notebook kernel you have to save the work (if you want changes you've made in cell inputs), close, and re-open.

I really like the model of the "direct connection" via the extension.  It's not that I mind the Python dependency so much--even though I've given up on PyPlot. I still have Python and performance of notebooks with Python runtime is fine. At this point it is a more reliable way to run Julia Notebooks.  

It's just that the extra features that are enabled are really nice--especially the variable viewer. And if the connections were reliable, it's just so nice to open a notebook and start running the cells. And I can see future features that might be provided like the debugger or converting notebooks to Julia script (like jupytext or Pluto) or creating Julia Markdown versions of the notebook.  i realize no promises, but it seems like these might be easier within VS Code using the rich Julia extension compared to doing these sorts of things in standalone Jupyter (Notebook or Lab).

In any case, this is a great direction. Let me know if there are more explicit things I can test to help diagnose this.

Also, I am on Apple Silicon with Julia 1.7 (so nice) and VS Code Apple Silicon build (also nice).

**Comment by ShuhuaGao** [comment]
The above issue  2 and 3 also exist on Windows 10 with Julia 1.7.

**Comment by lewisl** [comment]
Some improvement here. 

I changed settings for the built-in notebook support in VS Code.  One of the changes eliminated the frequent terminations of the Julia kernel. Alas, I changed 5 settings at once so I can't pinpoint which one solved the problem.

From my Julia Discourse post on the topic: 
```
"jupyter.experiments.enabled": false,
"jupyter.jupyterInterruptTimeout": 50000,
"jupyter.jupyterLaunchTimeout": 20000,
"jupyter.jupyterServerType": "local",
"jupyter.useDefaultConfigForJupyter": false,
```

**Comment by davidanthoff** [comment]
It looks to me as if you are using the Jupyter extension to communicate with the Julia kernels, in which case none of the features in the Julia extension even come into play at all.

**Comment by lewisl** [comment]
of course you are right. I can reliably use the vscode plugin’s notebook support.

- Lewis

On Jun 1, 2022, at 2:25 PM, David Anthoff ***@***.***> wrote:

﻿

It looks to me as if you are using the Jupyter extension to communicate with the Julia kernels, in which case none of the features in the Julia extension even come into play at all.

**Comment by lewisl** [comment]
I have to continue to report instability.

The vscode internal notebook will often hang when using the kernel provided by the Julia vscode extension.

In particular, Plots slow compilation time will cause the kernel to hang. 

When using Jupyter, running it from a command line, and opening the same notebook, the cell that runs Plots does not hang and generally compiles and produces the plot within 5-7 seconds.

**Comment by pfitzseb** [comment]
> Even quitting and restarting VS Code doesn't fix it. Only cold booting the Mac might enable it to work.

It's unclear what this means. We definitely don't have any logic to reconnect to notebook kernels after VS Code was closed and re-opened.

**Comment by lewisl** [comment]
Not that hard.  What it means is that the notebook controls to stop or restart the notebook don’t work.

Further, quitting and restarting VS Code doesn’t resolve the instability: a few cells will run and then the notebook will hang again.

If it helps, the cell that always hangs runs the plot command from the Plots package. This points to the possibility that compilation of the Plots package when run for the first time seems to cause the hang. (This is new information not in my prior posts.)

Only cold booting the machine will enable VS Code to open and run the notebook in a stable state. After a number of executions of the cells, the instability returns.

Sorry, if I was too terse in my prior description. I can’t fix this on my own—I am willing to run more focused cases to try to isolate the problem.

From: Sebastian Pfitzner ***@***.***>
Reply-To: julia-vscode/julia-vscode ***@***.***>
Date: Friday, June 3, 2022 at 12:40 AM
To: julia-vscode/julia-vscode ***@***.***>
Cc: Lewis Levin ***@***.***>, State change ***@***.***>
Subject: Re: [julia-vscode/julia-vscode] quite a few problems with built-in Notebook support (Issue #2638)


Even quitting and restarting VS Code doesn't fix it. Only cold booting the Mac might enable it to work.

It's unclear what this means. We definitely don't have any logic to reconnect to notebook kernels after VS Code was closed and re-opened.

**Comment by lewisl** [comment]
Until the issues are resolved with either VS Code's internal notebook support--likely the culprit--or the 'kernel' provided by the extension, I'll just stick to Jupyter Lab in a browser: stable and reliable.

The benefits aren't that great and there are too many limitations:
- few if any Jupyter extensions like nbconvert and jupytext
- WebIO doesn't work so no TableView support

The benefits are:
- all of VS Code's editing features
- nice contextual controls

But, not if it hangs.

**Comment by lewisl** [comment]
I am going to re-close this with the updated information.

I reported similar problems on the IJulia repo.

The short of it (skipping the long of it) is that it's not any of timeouts, insufficient memory, or which Plots backend is used.

It is more likely a mix of Apple Silicon code (Julia beta 1.7.2, VS Code) and Intel code running on Apple's intel emulator Rosetta.

I deleted all of my Julia installation (Julia + packages) and reinstalled everything on latest Julia stable 1.7.3 (Intel + Rosetta).

No problems.  Very hard stuff to diagnose.
